src/t8_schemes/t8_new_feature/t8_subelements/t8_subelements_quad_cxx.hxx

```diff
@@ -0,0 +1,357 @@
+
+  You should have received a copy of the GNU General Public License
+  along with t8code; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/** \file t8_default_quad.h
+ * We use a p4est_quadrant_t object as storage for the T8 quadrant.
+ * To record if and if yes, how this quadrant is part of a 3D octant, we use
+ * the member pad8 for the surrounding toplevel dimension (2 or 3), pad16 for
+ * the direction of its normal relative to a toplevel octant (0, 1, or 2), and
+ * p.user_long for the p4est_qcoord_t coordinate in the normal direction.
+ */
+
+#ifndef T8_DEFAULT_QUAD_CXX_HXX
+#define T8_DEFAULT_QUAD_CXX_HXX
+
+#include <p4est.h>
+#include <t8_element_cxx.hxx>
+
+#include "t8_schemes/t8_default/t8_default_line_cxx.hxx"
+#include "t8_schemes/t8_default/t8_default_common_cxx.hxx"
+
+/** The structure holding a quadrilateral element in the default scheme.
+ * We make this definition public for interoperability of element classes.
+ * We might want to put this into a private, scheme-specific header file.
+ */
+
+/* Define the struct, that stores all information needed for the quad scheme and subelements.
+ * 
+ *         p4est quadrant        recursive quad       refinement, using a transition 
+ *                                 refinement             cell with subelements
+ *         x - - - - - x         x - - x - - x                x - - - - - x
+ *         |           |         |     |     |                | \   2   / |       
+ *         |           |         |     |     |                | 1 \   /   |
+ *         |           |   -->   x - - x - - x       or       x - - x   3 |
+ *         |           |         |     |     |                | 0 / | \   |
+ *         |           |         |     |     |                | / 5 | 4 \ |
+ *         x - - - - - x         x - - x - - x                x - - x - - x 
+ * 
+ * A p4est quadrant can be refined, using either the standard quad scheme, or a transition cell, consisting of different subelements. 
+ * The quad refinement scheme is recursive, whereas a transition cell can only be used once, for example to remove hanging nodes, after the mesh has been adapted and balanced. 
+ * There are different types of transition cells possible, which we will refer to as subelement_type. 
+ * Each transition cell consists of different subelements. The given example consists of 6 different subelements, whose ids range from 0 to 5.
+ * A dummy variable will store the information, whether a given element is a subelement or a standard quad element. */ 
+
+typedef struct
+{
+  p4est_quadrant_t    p4q;                    /* p4est quadrant */                                                  
+  int                 dummy_is_subelement;    /* is a given element a subelement? */                                
+  int                 subelement_type;        /* which type of subelement to use? */                                 
```

Comment:
Is the type required as information?

Feels like once the subelement is constructed its behavior is independent of the type?
I.e. two subelements can have the same shape but differ only because they came initially from a different type. Why should these subelements be considered different? Would make computing coords etc. simpler if they were just the same subelement.

Reply:
ahh, okay so type + id is the unique identifier here